libraries/bot-builder/src/main/java/com/microsoft/bot/builder/CloudAdapterBase.java

```diff
@@ -0,0 +1,456 @@
+                    response = null;
+                } else if (
+                    activity.isType(ActivityTypes.TRACE)
+                        && !StringUtils.equals(activity.getChannelId(), Channels.EMULATOR)
+                ) {
+                    // no-op
+                    response = null;
+                } else if (StringUtils.isNotBlank(activity.getReplyToId())) {
+                    ConnectorClient connectorClient = context.getTurnState().get(BotFrameworkAdapter.CONNECTOR_CLIENT_KEY);
+                    response = connectorClient.getConversations().replyToActivity(activity).join();
+                } else {
+                    ConnectorClient connectorClient = context.getTurnState().get(BotFrameworkAdapter.CONNECTOR_CLIENT_KEY);
+                    response = connectorClient.getConversations().sendToConversation(activity).join();
+                }
+                if (response == null) {
+                    response = new ResourceResponse((activity.getId() == null) ? "" : activity.getId());
+                }
+
+                responses[index] = response;
+            }
+
+            return responses;
+        }, ExecutorFactory.getExecutor());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<ResourceResponse> updateActivity(TurnContext context, Activity activity) {
+        if (context == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("context"));
+        }
+
+        if (activity == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("activity"));
+        }
+
+        logger.info(String.format("updateActivity activityId: %d", activity.getId()));
+
+        ConnectorClient connectorClient = context.getTurnState().get(BotFrameworkAdapter.CONNECTOR_CLIENT_KEY);
+        return connectorClient.getConversations().updateActivity(activity);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<Void> deleteActivity(TurnContext context, ConversationReference reference) {
+        if (context == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("context"));
+        }
+
+        if (reference == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("reference"));
+        }
+
+        logger.info(String.format("deleteActivity Conversation id: %d, activityId: %d",
+            reference.getConversation().getId(),
+            activity.getId()));
+
+        ConnectorClient connectorClient = context.getTurnState().get(BotFrameworkAdapter.CONNECTOR_CLIENT_KEY);
+        return connectorClient.getConversations()
+            .deleteActivity(reference.getConversation().getId(), reference.getActivityId());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<Void> continueConversation(
+        String botAppId,
```

Comment:
I suppose there's a good reason but why this is a string and not a claimsIdentity?

Reply:
my bad, confused with the other overload